aptos-move/aptos-gas-algebra/src/abstract_algebra.rs

```diff
@@ -0,0 +1,272 @@
+ * Gas Expression & Visitor
+ *
+ **************************************************************************************************/
+/// Trait representing an abstract view over an expression that is used to represent some
+/// gas amount.
+///
+/// It carries a type parameter `E`, indicating an environment in which the expression can be
+/// evaluated/materialized.
+pub trait GasExpression<E> {
+    type Unit;
+
+    fn evaluate(&self, feature_version: u64, env: &E) -> GasQuantity<Self::Unit>;
+
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor);
+
+    fn per<U>(self) -> GasPerUnit<Self, U>
+    where
+        Self: Sized,
+    {
+        GasPerUnit {
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An interface for performing post-order traversal of the tree structure of a gas expression.
+///
+/// Alternatively, one could think that the callbacks are invoked following the Reverse Polish
+/// notation of the expression.
+///
+/// Here are a few examples:
+/// - `1 + 2`
+///   - `quantity(1)`
+///   - `quantity(2)`
+///   - `add()`
+/// - `A + B * 50`
+///   - `gas_param<A>()`
+///   - `gas_param<B>()`
+///   - `quantity(50)`
+///   - `mul()`
+///   - `add()`
+pub trait GasExpressionVisitor {
+    fn add(&mut self);
+
+    fn mul(&mut self);
+
+    fn gas_param<P>(&mut self);
+
+    fn quantity<U>(&mut self, quantity: GasQuantity<U>);
+
+    fn per<U>(&mut self);
+}
+
+/***************************************************************************************************
+ * Built-in Gas Expressions
+ *
+ **************************************************************************************************/
+#[derive(Debug, Clone)]
+pub struct GasAdd<L, R> {
+    pub left: L,
+    pub right: R,
+}
+
+#[derive(Debug, Clone)]
+pub struct GasMul<L, R> {
+    pub left: L,
+    pub right: R,
+}
+
+#[derive(Debug, Clone)]
+pub struct GasPerUnit<T, U> {
+    pub inner: T,
+    phantom: PhantomData<U>,
+}
+
+/***************************************************************************************************
+ * Gas Expression Impl
+ *
+ **************************************************************************************************/
+impl<E, T> GasExpression<E> for &T
+where
+    T: GasExpression<E>,
+{
+    type Unit = T::Unit;
+
+    fn evaluate(&self, feature_version: u64, env: &E) -> GasQuantity<Self::Unit> {
+        (*self).evaluate(feature_version, env)
+    }
+
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor) {
+        (*self).visit(visitor)
+    }
+}
+
+impl<E, U> GasExpression<E> for GasQuantity<U> {
+    type Unit = U;
+
+    fn evaluate(&self, _feature_version: u64, _env: &E) -> GasQuantity<Self::Unit> {
+        *self
+    }
+
+    fn visit(&self, visitor: &mut impl GasExpressionVisitor) {
+        visitor.quantity(*self)
+    }
+}
+
+impl<E, L, R, U> GasExpression<E> for GasAdd<L, R>
```

Comment:
I have the sinking feeling that I have a very different view on good code then everybody else here. I'm a big believer in typing and leveraging typing to maintain invariants, but typing also comes with costs. More code, more symbols/names you are polluting your mental ram with, boilerplate, restriction of functions accept them. The art of coding is to find the balance between leveraging static typing and maintaining information/invariants dynamically, such to produce efficient, readable and short/dense code.

All this code here generating a whole expression tree, where I think in the end the gas calculation is just inner_product(gas_params, size_vector). I feel we are overusing the typesystem for what is basic linear algebra. Note that in physics matrix entries often have different units, but in physics libraries you always store them as just numbers in arrays. The ability to operate on matrixes dynamically (looping) trumps the benefit of type safety that comes from encoding units in the typesystem.

Reply:
BTW. Taking the perspective of linear algebra also makes obvious a no-code solution. If we have some benchmark suite of tx's for which we know the desired cost. Let's name them t_i and c_i for i = 1...#benchmarks, then for each gas param g_j (j = 1 .. #gas params) we can calculate the gas of the transaction where we set g_j = 1 all other params 0. This gives a matrix C_{i j} for which it holds,  C.g is the vector of the gas costs of each benchmark when the parameters are set by g. So that means we want to chose g such that C.g \approx c by using something like least square.

Reply:
> So that means we want to chose g such that C.g \approx c by using something like least square.

This is exactly the idea behind the automated gas calibration tool @wlawt and I are building ;)

Also thanks for the great write up! I think you started a very good discussion about the tradeoffs of having additional type safety. I'll chime in at little bit later and explain my views.

Reply:
My apologies for the long write-up below, but you brought up a great topic and I think it’s worth discussing it in depth.

At a high level, I totally agree with you that one aspect of coding is to balance between achieving the desired level of security/correctness and providing straightforward user interfaces. In some rare cases you could have both, but more likely you need to sacrifice one for the other to some extent, and the choice is often determined by the nature of the project you are working with.

### On Security & Correctness

Maybe I'm being a bit too paranoid, but I would argue that **as a blockchain platform, we MUST take correctness & security in the most serious fashion, more so than almost all other commercial applications**. A mistake on our platform could not only result in people losing assets, but also hard or even impossible to revert.

Now coming back to the gas algebra (both concrete & abstract), having the unit embedded into the types of the numbers does allow us to catch certain classes of bugs at compile time. Here is one of the simplest examples

- If the gas parameters are just treated as normal integers, then it's totally possible for one to miswrite `base_cost + unit_cost * len` as `unit_cost + base_cost * len` and having this bug hiding in the system for months without being detected.
- If unit annotations are required, then the compiler would be able to see that the multiplication is invalid — you can’t multiply `gas` & `byte`. (Well, units like `gas * byte` are common in physics but they don’t make much sense in our context.) Similarly, it would prevent one from adding `base_cost` and `unit_cost`.
- Alternatively, you could see the unit annotations as a kind of redundancy — it composes a proof that your gas calculation is valid. The current approach also does it in a way with a nice property: **the code is both the instruction for execution and the proof itself, not requiring separate specs.**

One could argue that these are probably not the most serious bugs we could have, but still, I’d rather have us ruling them out completely with the help of the type system, than having to deal the aftermath one day. This is particularly appealing if you see it from the gas metering system’s perspective — there will be more and more of us interfacing with the system and one day someone will inevitably make a mistake.

You also brought up another interesting scenario: scientific computing/physics simulation. IMO this has totally different security & correctness requirements. This kind of number crunching happens on the scientist’s own computer/server and there are other ways to detect errors and redo the experiment. Things are different if you are controlling a plane, but in those cases, formal verification (even more proof) is often a hard requirement. The upcoming gas calibrator will actually take a similar approach, erasing the type info and put the numbers into matrices.

### What About User Experience & Maintainability?

Now I’m aware that this additional complexity makes it harder for clients to use the gas metering system — because the gas quantities can have different types, you would need some sort of heterogeneous collection if you want to store them conveniently. Otherwise, you would have to require the users to break down their formulae into individual terms and charge them separately:

```rust
context.charge(PARAM_A, 1)?;
context.charge(PARAM_B, len)?
..
```

My current solution to this problem is to offer such a heterogeneous data structure (which conveniently takes the form of an expression), and hide the complexities of the generation logic from the users. I would claim that this does result in a workflow is intuitive enough to follow — for each new gas parameter, you simply add a line to declare it, and then you can use it in an expression.

```rust
define_gas_params!(
  ..  
  [base: InternalGas, "base", 100],
  [per_byte: InternalGasPerByte, "per_byte", 10],
  ..
)

context.charge(BASE + PER_BYTE * len)?;
```

The user code here is actually rather dense and easy to read. The compiler will effectively expand this into (a more efficient version of) the following code

```rust
context.charge(BASE, 1)?;
context.charge(PER_BYTE , len)?
```

I do understand that some people, myself included, don’t like to be told to simply “do this, do that and then you’ll be fine”, and have the desire to fully understand the esoteric signature of an API before feeling comfortable about calling it. 

I don’t think I have a general solution to this, but at least I think it can be somewhat mitigated by 
1. Defining a clear contract between components so you can compartmentalize your mental memory and not having to keep all the boilerplate & restrictions in mind all the time
2. Having good documentation explaining not just the inter-component APIs, but also the inner workings of the system, so a curious reader could eliminate the sense of uncertainty they feel

This is also related to another issue I want to touch on: maintainability. To compensate for the user experience above, I had to sacrifice the overall complexity of the system. I do recognize it being a rather unpleasant tradeoff, and I don’t know the problem well enough to give a general solution, but at least the complexity is on the library writer’s side, not the users’. I hope some of the simplifications I made (e.g. unifying the two generation macros into one) and comments I’m planning to add will make it easier for other people to understand and maintain this part of the code. (I don’t know if it counts as a relief but we’re not at Haskell’s level of typestate programming yet…)